rom/laghos_eqp.cpp

```diff
@@ -0,0 +1,1102 @@
+    MFEM_VERIFY(nsets == snapX.size() && nsets == snapV.size() &&
+                nsets == snapE.size() && nsets > 0, "");
+
+    Array<int> numSnapVar(3);
+    std::vector<Array<int>> allNumSnapVar(3);
+    for (int i=0; i<3; ++i)
+        allNumSnapVar[i].SetSize(nsets);
+
+    Array<int> allnsnap(nsets);
+
+    for (int i=0; i<nsets; ++i)
+    {
+        allNumSnapVar[0][i] = snapX[i]->numColumns();
+        allNumSnapVar[1][i] = snapV[i]->numColumns();
+        allNumSnapVar[2][i] = snapE[i]->numColumns();
+
+        // TODO: is there a better way?
+        allnsnap[i] = allNumSnapVar[0][i];
+        for (int j=1; j<3; ++j)
+        {
+            allnsnap[i] = std::max(allnsnap[i], allNumSnapVar[j][i]);
+        }
+    }
+
+    for (int i=0; i<3; ++i)
+        numSnapVar[i] = allNumSnapVar[i].Sum();
+
+    const int nsnap = numSnapVar.Max();
+
+    std::vector<Array<int>> numSkipped(3);
+    for (int i=0; i<3; ++i)
+    {
+        numSkipped[i].SetSize(nsets);
+
+        for (int j=0; j<nsets; ++j)
+        {
+            numSkipped[i][j] = allnsnap[j] - allNumSnapVar[i][j];
+        }
+
+        MFEM_VERIFY(numSkipped[i].Max() <= 1, "");
+    }
+
+    Vector r(nqe);
+
+    const int Gnsnap = input.window == 0 ? nsnap + 1 : nsnap;
+
+    // Compute G of size (NB * Gnsnap) x NQ, storing its transpose Gt.
+    CAROM::Matrix Gt(NQ, NB * Gnsnap, true);
+    cout << "NNLS using " << NB << " basis dim for equation E " << equationE
+         << " and " << nsnap << " snapshots" << endl;
+
+    // For 0 <= j < NB, 0 <= i <= nsnap, 0 <= e < ne, 0 <= m < nqe,
+    // G(j + (i*NB), (e*nqe) + m)
+    // is the coefficient of e_j^T M_e^{-1} F(v_i,e_i,x_i)^T v_i at point m of
+    // element e, with respect to the integration rule weight at that point,
+    // where the "exact" quadrature solution is ir0->GetWeights().
+
+    Vector v_i(tH1size);
+    Vector x_i(tH1size);
+    Vector e_i(tL2size);
+    Vector g_i(tL2size);
+
+    Vector w_j_e, v_i_e, v_j_e;
+
+    Vector S((2*input.H1FESpace->GetVSize()) + input.L2FESpace->GetVSize());
+    Vector S_v(S, input.H1FESpace->GetVSize(), input.H1FESpace->GetVSize());  // Subvector
+
+    MFEM_VERIFY(tH1size == basisV->numRows(), "");
+    MFEM_VERIFY(tL2size == basisE->numRows(), "");
+    CAROM::Matrix W(equationE ? L2size : H1size, NB, true);
+
+    ParGridFunction gf2H1(*gfH1);
+
+    for (int j=0; j<NB; ++j)
+    {
+        if (equationE)
+        {
+            for (int i=0; i<tL2size; ++i)
+                e_i[i] = (*basisE)(i,j);
+
+#ifdef EQP_NO_Minv
+            for (int i=0; i<tL2size; ++i)
+                W(i,j) = e_i[i];  // TODO: just set to basisE?
+#else
+            input.FOMoper->MultMeInv(e_i, g_i);
+
+            for (int i=0; i<tL2size; ++i)
+                W(i,j) = g_i[i];
+#endif
+        }
+        else
+        {
+            for (int i=0; i<tH1size; ++i)
+                v_i[i] = (*basisV)(i,j);
+
+            gfH1->SetFromTrueDofs(v_i);
+#ifdef EQP_NO_Minv
```

Comment:
Usually we select options with runtime flags, and that could be done here. In this case, the option to not use mass matrix inverses at all gives very bad results, and there are no cases in which this is known to be a good option. Some possibilities are

1. Remove the code in `#ifdef EQP_NO_Minv`, since it should not be used.
2. Add a runtime flag to choose whether to use mass matrix inverses.
3. Keep the code, so anyone curious can easily verify that it is a bad option.

I chose 3, just for easy testing in the future, without indicating that it is intended as a normal option.